File: PYTHON/university.py
# university.py
# Kelas University menampung array of object untuk Student, Lecturer, Staff, dan TA
# serta menyediakan fungsi untuk menampilkan tabel dinamis

from student import Student
from lecturer import Lecturer
from staff import Staff
from teaching_assistant import TeachingAssistant

class University:
    def __init__(self):
        self.students = []
        self.lecturers = []
        self.staffs = []
        self.tas = []

    def add_student(self, s: Student): self.students.append(s)
    def add_lecturer(self, l: Lecturer): self.lecturers.append(l)
    def add_staff(self, st: Staff): self.staffs.append(st)
    def add_ta(self, t: TeachingAssistant): self.tas.append(t)

    # Helper untuk membuat garis tabel
    def __line(self, width):
        print("+" + "+".join(["-" * (w + 2) for w in width]) + "+")

    # Helper untuk cetak tabel generik
    def __print_table(self, header, rows):
        width = [len(h) for h in header]
        for row in rows:
            for i, col in enumerate(row):
                width[i] = max(width[i], len(str(col)))

        self.__line(width)
        print("|" + "|".join([f" {header[i]:<{width[i]}} " for i in range(len(header))]) + "|")
        self.__line(width)
        for row in rows:
            print("|" + "|".join([f" {str(row[i]):<{width[i]}} " for i in range(len(row))]) + "|")
        self.__line(width)

    def print_students(self):
        print("\n=== DATA STUDENT ===")
        rows = [[s.get_id(), s.get_nama(), s.get_usia(),
                 s.get_nim(), s.get_jurusan(), s.get_gpa()] for s in self.students]
        self.__print_table(["ID","Nama","Usia","NIM","Jurusan","GPA"], rows)

    def print_lecturers(self):
        print("\n=== DATA LECTURER ===")
        rows = [[d.get_id(), d.get_nama(), d.get_usia(),
                 d.get_nidn(), d.get_departemen(), d.get_jabatan()] for d in self.lecturers]
        self.__print_table(["ID","Nama","Usia","NIDN","Departemen","Jabatan"], rows)

    def print_staffs(self):
        print("\n=== DATA STAFF ===")
        rows = [[s.get_id(), s.get_nama(), s.get_usia(),
                 s.get_posisi(), int(s.get_gaji()), s.get_shift()] for s in self.staffs]
        self.__print_table(["ID","Nama","Usia","Posisi","Gaji","Shift"], rows)

    def print_tas(self):
        print("\n=== DATA TEACHING ASSISTANT ===")
        rows = [[t.get_id(), t.get_nama(), t.get_usia(),
                 t.get_nim(), t.get_jurusan(), t.get_gpa(),
                 t.get_jam_asistensi(), t.get_mata_kuliah()] for t in self.tas]
        self.__print_table(["ID","Nama","Usia","NIM","Jurusan","GPA","JamAsistensi","MataKuliah"], rows)

File: JAVA/University.java
// University.java
// Menampung array of object Student, Lecturer, Staff, dan TA
// serta mencetak data dengan format tabel sederhana

import java.util.*;

// University class
public class University {
    private List<Student> students = new ArrayList<>();
    private List<Lecturer> lecturers = new ArrayList<>();
    private List<Staff> staffs = new ArrayList<>();
    private List<TeachingAssistant> tas = new ArrayList<>();

    // ---------------- Helper ----------------
    private void printLine(int[] width) {
        for (int w : width) {
            System.out.print("+");
            for (int j = 0; j < w + 2; j++) System.out.print("-");
        }
        System.out.println("+");
    }

    private void printTable(String[] header, List<String[]> rows) {
        int cols = header.length;
        int[] width = new int[cols];

        // inisialisasi dengan panjang header
        for (int i = 0; i < cols; i++) width[i] = header[i].length();

        // cek panjang maksimum isi data
        for (String[] row : rows) {
            for (int i = 0; i < cols; i++) {
                if (row[i].length() > width[i]) width[i] = row[i].length();
            }
        }

        // cetak header
        printLine(width);
        System.out.print("|");
        for (int i = 0; i < cols; i++) {
            System.out.print(" " + header[i] + " ".repeat(width[i] - header[i].length() + 1) + "|");
        }
        System.out.println();
        printLine(width);

        // cetak isi
        for (String[] row : rows) {
            System.out.print("|");
            for (int i = 0; i < cols; i++) {
                System.out.print(" " + row[i] + " ".repeat(width[i] - row[i].length() + 1) + "|");
            }
            System.out.println();
        }
        printLine(width);
    }

    // ---------------- Adders ----------------
    public void addStudent(Student s) { students.add(s); }
    public void addLecturer(Lecturer l) { lecturers.add(l); }
    public void addStaff(Staff st) { staffs.add(st); }
    public void addTA(TeachingAssistant t) { tas.add(t); }

    // ---------------- Printers ----------------
    public void printStudents() {
        System.out.println("\n=== DATA STUDENT ===");
        String[] header = {"ID","Nama","Usia","NIM","Jurusan","GPA"};
        List<String[]> rows = new ArrayList<>();
        for (Student s : students) {
            rows.add(new String[]{
                s.getId(),
                s.getNama(),
                String.valueOf(s.getUsia()),
                s.getNim(),
                s.getJurusan(),
                String.format("%.2f", s.getGpa())
            });
        }
        printTable(header, rows);
    }

    public void printLecturers() {
        System.out.println("\n=== DATA LECTURER ===");
        String[] header = {"ID","Nama","Usia","NIDN","Departemen","Jabatan"};
        List<String[]> rows = new ArrayList<>();
        for (Lecturer d : lecturers) {
            rows.add(new String[]{
                d.getId(),
                d.getNama(),
                String.valueOf(d.getUsia()),
                d.getNidn(),
                d.getDepartemen(),
                d.getJabatan()
            });
        }
        printTable(header, rows);
    }

    public void printStaffs() {
        System.out.println("\n=== DATA STAFF ===");
        String[] header = {"ID","Nama","Usia","Posisi","Gaji","Shift"};
        List<String[]> rows = new ArrayList<>();
        for (Staff st : staffs) {
            rows.add(new String[]{
                st.getId(),
                st.getNama(),
                String.valueOf(st.getUsia()),
                st.getPosisi(),
                String.valueOf((int) st.getGaji()),
                st.getShift()
            });
        }
        printTable(header, rows);
    }

    public void printTAs() {
        System.out.println("\n=== DATA TEACHING ASSISTANT ===");
        String[] header = {"ID","Nama","Usia","NIM","Jurusan","GPA","JamAsistensi","MataKuliah"};
        List<String[]> rows = new ArrayList<>();
        for (TeachingAssistant t : tas) {
            rows.add(new String[]{
                t.getId(),
                t.getNama(),
                String.valueOf(t.getUsia()),
                t.getNim(),
                t.getJurusan(),
                String.format("%.2f", t.getGpa()),
                String.valueOf(t.getJamAsistensi()),
                t.getMataKuliah()
            });
        }
        printTable(header, rows);
    }
}